### ISSUE TITLE
v8 runtime crashed when creating lots of wasm vms

### ISSUE THREAD

**Issue by johnlanni** [envoyproxy/envoy#19720]
*Description*:
v8 runtime crashed when creating lots of wasm vms.

*Repro steps*:
envoy version: 1.21.0
I enabled 5 wasm filters with different wasm file, and run envoy with `--concurrency 50`

*Config*:
```yaml
node:
  cluster: test-cluster
  id: test

admin:
  address:
    socket_address: { address: 127.0.0.1, port_value: 9901 }

static_resources:
  listeners:
  - name: listener_0
    address:
      socket_address: { address: 127.0.0.1, port_value: 10000 }
    filter_chains:
    - filters:
      - name: envoy.filters.network.http_connection_manager
        typed_config:
          "@type": type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager
          stat_prefix: ingress_http
          codec_type: AUTO
          route_config:
            name: local_route
            virtual_hosts:
            - name: local_service
              domains: ["*"]
              routes:
              - match: { prefix: "/" }
                route: { cluster: some_service }
          http_filters:
          - name: wasmtest
            typed_config:
              "@type": type.googleapis.com/udpa.type.v1.TypedStruct
              type_url: type.googleapis.com/envoy.extensions.filters.http.wasm.v3.Wasm
              value:
                config:
                  name: w1
                  vm_config:
                    runtime: envoy.wasm.runtime.v8
                    code:
                      local:
                        filename: /tmp/1.wasm
                    allow_precompiled: true
          - name: wasmtest
            typed_config:
              "@type": type.googleapis.com/udpa.type.v1.TypedStruct
              type_url: type.googleapis.com/envoy.extensions.filters.http.wasm.v3.Wasm
              value:
                config:
                  name: w2
                  vm_config:
                    runtime: envoy.wasm.runtime.v8
                    code:
                      local:
                        filename: /tmp/2.wasm
                    allow_precompiled: true
          - name: wasmtest
            typed_config:
              "@type": type.googleapis.com/udpa.type.v1.TypedStruct
              type_url: type.googleapis.com/envoy.extensions.filters.http.wasm.v3.Wasm
              value:
                config:
                  name: w3
                  vm_config:
                    runtime: envoy.wasm.runtime.v8
                    code:
                      local:
                        filename: /tmp/3.wasm
                    allow_precompiled: true
          - name: wasmtest
            typed_config:
              "@type": type.googleapis.com/udpa.type.v1.TypedStruct
              type_url: type.googleapis.com/envoy.extensions.filters.http.wasm.v3.Wasm
              value:
                config:
                  name: w4
                  vm_config:
                    runtime: envoy.wasm.runtime.v8
                    code:
                      local:
                        filename: /tmp/4.wasm
                    allow_precompiled: true
          - name: wasmtest
            typed_config:
              "@type": type.googleapis.com/udpa.type.v1.TypedStruct
              type_url: type.googleapis.com/envoy.extensions.filters.http.wasm.v3.Wasm
              value:
                config:
                  name: w5
                  vm_config:
                    runtime: envoy.wasm.runtime.v8
                    code:
                      local:
                        filename: /tmp/5.wasm
                    allow_precompiled: true
          - name: envoy.filters.http.router
  clusters:
  - name: some_service
    connect_timeout: 0.25s
    type: STATIC
    lb_policy: ROUND_ROBIN
    load_assignment:
      cluster_name: some_service
      endpoints:
      - lb_endpoints:
        - endpoint:
            address:
              socket_address:
                address: 127.0.0.1
                port_value: 12345
```

*Logs*:
last few lines of log:
```bash
<--- Last few GCs --->


<--- JS stacktrace --->


#
# Fatal javascript OOM in GC during deserialization
#

Trace/breakpoint trap
```

*Call Stack*:

coredump backtrace:
```bash
#0  0x000055d6b6739226 in operator() () at ../../src/base/platform/platform-posix.cc:571
#1  v8::base::OS::Abort () at ../../src/base/platform/platform-posix.cc:571
#2  0x000055d6b673eb82 in v8::Utils::ReportOOMFailure () at ../../src/api/api.cc:372
#3  v8::Utils::ReportOOMFailure () at ../../src/api/api.cc:359
#4  0x000055d6b673ee2b in v8::internal::V8::FatalProcessOutOfMemory () at ../../src/api/api.cc:338
#5  0x000055d6b6809fd9 in v8::internal::Heap::FatalProcessOutOfMemory () at ../../src/heap/heap.cc:6286
#6  0x000055d6b681dc84 in v8::internal::Heap::CollectGarbage () at ../../src/heap/heap.cc:1674
#7  0x000055d6b681fd42 in v8::internal::Heap::AllocateRawWithLightRetrySlowPath () at ../../src/heap/heap.cc:5547
#8  0x000055d6b681fdb8 in v8::internal::Heap::AllocateRawWithRetryOrFailSlowPath () at ../../src/heap/heap.cc:5564
#9  0x000055d6b6ad641d in v8::internal::Heap::AllocateRawWith<(v8::internal::Heap::AllocationRetryMode)1> () at ../../src/heap/heap-inl.h:344
#10 v8::internal::Heap::AllocateRawOrFail () at ../../src/heap/heap-inl.h:353
#11 v8::internal::Deserializer<v8::internal::Isolate>::Allocate () at ../../src/snapshot/deserializer.cc:1268
#12 0x000055d6b6adad21 in v8::internal::Deserializer<v8::internal::Isolate>::ReadObject () at ../../src/snapshot/deserializer.cc:639
#13 0x000055d6b6adb306 in v8::internal::Deserializer<v8::internal::Isolate>::ReadSingleBytecodeData<v8::internal::SlotAccessorForRootSlots> ()
    at ../../src/snapshot/deserializer.cc:915
#14 v8::internal::Deserializer<v8::internal::Isolate>::ReadData () at ../../src/snapshot/deserializer.cc:897
#15 0x000055d6b6808c6c in v8::internal::Heap::IterateRoots () at ../../src/heap/heap.cc:4824
#16 0x000055d6b6ae7cd2 in v8::internal::StartupDeserializer::DeserializeIntoIsolate () at ../../src/snapshot/startup-deserializer.cc:31
#17 0x000055d6b67c8ff4 in v8::internal::Isolate::Init () at ../../src/execution/isolate.cc:3942
#18 0x000055d6b6ae529c in v8::internal::Snapshot::Initialize () at ../../src/snapshot/snapshot.cc:194
#19 0x000055d6b6747197 in v8::Isolate::Initialize () at ../../src/api/api.cc:8674
#20 0x000055d6b67474f2 in v8::Isolate::New () at ../../src/api/api.cc:8721
#21 0x000055d6b6732655 in wasm::Store::make () at ../../src/wasm/c-api.cc:487
#22 0x000055d6b672c84e in proxy_wasm::(anonymous namespace)::V8::clone() () at external/proxy_wasm_cpp_host/src/v8/v8.cc:282
#23 0x000055d6b670139c in proxy_wasm::WasmBase::WasmBase(std::shared_ptr<proxy_wasm::WasmHandleBase> const&, std::function<std::unique_ptr<proxy_wasm::WasmVm, std::default_delete<proxy_wasm::WasmVm> > ()>) () at /usr/include/c++/9/bits/unique_ptr.h:360
#24 0x000055d6b6466fa9 in Envoy::Extensions::Common::Wasm::Wasm::Wasm(std::shared_ptr<Envoy::Extensions::Common::Wasm::WasmHandle>, Envoy::Event::Dispatcher&) ()
    at /usr/include/c++/9/ext/atomicity.h:96
#25 0x000055d6b6467572 in std::_Function_handler<std::shared_ptr<proxy_wasm::WasmHandleBase> (std::shared_ptr<proxy_wasm::WasmHandleBase>), Envoy::Extensions::Common::Wasm::getWasmHandleCloneFactory(Envoy::Event::Dispatcher&, std::function<proxy_wasm::ContextBase* (Envoy::Extensions::Common::Wasm::Wasm*, std::shared_ptr<Envoy::Extensions::Common::Wasm::Plugin> const&)>)::{lambda(std::shared_ptr<proxy_wasm::WasmHandleBase>)#1}>::_M_invoke(std::_Any_data const&, std::shared_ptr<proxy_wasm::WasmHandleBase>&&) () at /usr/include/c++/9/new:174
#26 0x000055d6b6703e0f in proxy_wasm::getOrCreateThreadLocalPlugin(std::shared_ptr<proxy_wasm::WasmHandleBase>, std::shared_ptr<proxy_wasm::PluginBase>, std::function<std::shared_ptr<proxy_wasm::WasmHandleBase> (std::shared_ptr<proxy_wasm::WasmHandleBase>)>, std::function<std::shared_ptr<proxy_wasm::PluginHandleBase> (std::shared_ptr<proxy_wasm::WasmHandleBase>, std::shared_ptr<proxy_wasm::PluginBase>)>) () at /usr/include/c++/9/bits/std_function.h:683
#27 0x000055d6b64649f9 in Envoy::Extensions::Common::Wasm::getOrCreateThreadLocalPlugin(std::shared_ptr<Envoy::Extensions::Common::Wasm::WasmHandle> const&, std::shared_ptr<Envoy::Extensions::Common::Wasm::Plugin> const&, Envoy::Event::Dispatcher&, std::function<proxy_wasm::ContextBase* (Envoy::Extensions::Common::Wasm::Wasm*, std::shared_ptr<Envoy::Extensions::Common::Wasm::Plugin> const&)>) () at /usr/include/c++/9/bits/shared_ptr_base.h:756
#28 0x000055d6b623c129 in std::_Function_handler<std::shared_ptr<Envoy::Extensions::Common::Wasm::PluginHandleSharedPtrThreadLocal> (Envoy::Event::Dispatcher&), Envoy::Extensions::HttpFilters::Wasm::FilterConfig::FilterConfig(envoy::extensions::filters::http::wasm::v3::Wasm const&, Envoy::Server::Configuration::FactoryContext&)::{lambda(std::shared_ptr<Envoy::Extensions::Common::Wasm::WasmHandle> const&)#1}::operator()(std::shared_ptr<Envoy::Extensions::Common::Wasm::WasmHandle> co--Type <RET> for more, q to quit, c to continue without paging--
nst&) const::{lambda(Envoy::Event::Dispatcher&)#1}>::_M_invoke(std::_Any_data const&, Envoy::Event::Dispatcher&) () at /usr/include/c++/9/bits/std_function.h:403
warning: Could not find DWO CU bazel-out/k8-opt/bin/source/extensions/access_loggers/wasm/_objs/config/config.dwo(0x30382dba5652e41d) referenced by CU at offset 0x8bc [in module /mnt/zhangty/envoy-v8]
#29 0x000055d6b5ea77e5 in std::_Function_handler<std::shared_ptr<Envoy::ThreadLocal::ThreadLocalObject> (Envoy::Event::Dispatcher&), std::function<std::shared_ptr<Envoy::Extensions::Common::Wasm::PluginHandleSharedPtrThreadLocal> (Envoy::Event::Dispatcher&)> >::_M_invoke(std::_Any_data const&, Envoy::Event::Dispatcher&)
    () at /usr/include/c++/9/bits/std_function.h:683
warning: Could not find DWO CU bazel-out/k8-opt/bin/source/common/thread_local/_objs/thread_local_lib/thread_local_impl.dwo(0xb948ecb930b00216) referenced by CU at offset 0x1b60ee5 [in module /mnt/zhangty/envoy-v8]
#30 0x000055d6b74b7cff in std::_Function_handler<void (), Envoy::ThreadLocal::InstanceImpl::SlotImpl::set(std::function<std::shared_ptr<Envoy::ThreadLocal::ThreadLocalObject> (Envoy::Event::Dispatcher&)>)::{lambda()#1}>::_M_invoke(std::_Any_data const&) () at /usr/include/c++/9/bits/std_function.h:683
#31 0x000055d6b7c7dffa in Envoy::Event::DispatcherImpl::runPostCallbacks() () at /usr/include/c++/9/bits/std_function.h:683
#32 0x000055d6b7c7e0c0 in Envoy::Event::DispatcherImpl::run(Envoy::Event::Dispatcher::RunType) () at source/common/event/dispatcher_impl.cc:275
#33 0x000055d6b74d8c41 in Envoy::Server::WorkerImpl::threadRoutine(Envoy::Server::GuardDog&, std::function<void ()> const&) ()
    at /usr/include/c++/9/bits/unique_ptr.h:360
#34 0x000055d6b7fb08f9 in Envoy::Thread::ThreadImplPosix::ThreadImplPosix(std::function<void ()>, absl::optional<Envoy::Thread::Options> const&)::{lambda(void*)#1}::_FUN(void*) () at /usr/include/c++/9/bits/std_function.h:683
#35 0x00007f0f732d1609 in start_thread (arg=<optimized out>) at pthread_create.c:477
#36 0x00007f0f731f8263 in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```

**Comment by mattklein123** [comment]
cc @PiotrSikora

**Comment by PiotrSikora** [comment]
@johnlanni thanks for the report, fixed in #19763.
